eosio.unregd/eosio.unregd.cpp

```diff
@@ -21,6 +21,103 @@ void unregd::add(const ethereum_address& ethereum_address, const asset& balance)
   });
 }
 
+/**
+ * Register an EOS account using the stored information (address/balance) verifying an ETH signature
+ */
+void unregd::regaccount(const bytes& signature, const string& account, const eosio::public_key& eos_pubkey) {
+
+  eosio_assert(signature.size() == 66, "Invalid signature");
+  eosio_assert(account.size() == 12, "Invalid account length");
+
+  // Verify that the destination account name is valid
+  for(const auto& c : account) {
+    if(!((c >= 'a' && c <= 'z') || (c >= '1' && c <= '5')))
+      eosio_assert(false, "Invalid account name");
+  }
+
+  auto naccount = string_to_name(account.c_str());
+
+  // Verify that the account does not exists
+  eosio_assert(!is_account(naccount), "Account already exists");
+
+  // Calculate message hash based on current TX block num/prefix
+  char* message = (char*)malloc(64);
+  sprintf(message, "%u,%u", tapos_block_num(), tapos_block_prefix());
+
+  checksum256 digest;
+  sha256(message, strlen(message), &digest);
+
+  // Recover compressed pubkey from signature
+  uint8_t* pubkey = (uint8_t*)malloc(64);
+  uint8_t* compressed_pubkey = (uint8_t*)malloc(34);
+  auto res = recover_key(
+    &digest,
+    signature.data(),
+    signature.size(),
+    (char*)compressed_pubkey,
+    34
+  );
+
+  eosio_assert(res == 34, "Recover key failed");
+
+  // Decompress pubkey
+  uECC_decompress(compressed_pubkey+1, pubkey, uECC_secp256k1());
+
+  // Calculate ETH address based on decompressed pubkey
+  sha3_ctx* shactx = (sha3_ctx*)malloc(sizeof(sha3_ctx));
+
+  checksum256 msghash;
+  rhash_keccak_256_init(shactx);
+  rhash_keccak_update(shactx, pubkey, 64);
+  rhash_keccak_final(shactx, msghash.hash);
+
+  uint8_t* eth_address = (uint8_t*)malloc(20);
+  memcpy(eth_address, msghash.hash + 12, 20);
+
+  // Verify that the ETH address exists in the "addresses" eosio.unregd contract table
+  addresses_index addresses(_self, _self);
+  auto index = addresses.template get_index<N(ethereum_address)>();
+
+  auto itr = index.find(compute_ethereum_address_key256(eth_address));
+  eosio_assert(itr != index.end(), "Address not found");
+
+  // Split contribution balance into cpu/net/liquid
+  auto balances = split_snapshot_abp(itr->balance);
+  eosio_assert(balances.size() == 3, "Unable to split snapshot");
+  eosio_assert(itr->balance == balances[0] + balances[1] + balances[2], "internal error");
+
+  // Calculate the amount of EOS to purchase 8k of RAM
+  auto amount_to_purchase_8kb_of_RAM = buyrambytes(8*1024);
+
+  // Build authority with the pubkey passed as parameter
+  auto auth = authority{1,{{eos_pubkey,1}},{},{}};
+
+  // Issue to eosio.unregd the necesary EOS to buy 8K of RAM
+  INLINE_ACTION_SENDER(call::token, issue)( N(eosio.token), {{N(eosio),N(active)}},
```

Comment:
I feel uneasy that this contract is designed in a way that requires the `eosio.unregd` account to be privileged for it to be able to operate. It is best to reduce the attack surface of privileged code to just what really needs to be privileged (within additional reasonable constraints of performance requirements and maintainability).

As far as I can see, the only reasons that this contract needs to be deployed on a privileged account are:
1) So `eosio` can be the one that created the account (but why not just have `eosio.unregd` be the one to create the account?).
2) So that the contract can issue new EOS tokens into existence and use it to buy 8 KiB of RAM for the new account.

Regarding reason 2, I personally question the decision to allow the contract to issue into existence a practically unbounded amount of EOS (because the contract pays whatever it costs to buy 8 KiB of RAM regardless of what the price of RAM may be at the time). Nevertheless, if block producers do wish to continue with that policy, there are other ways of doing it that provides greater safeguards and allows the `eosio.unregd` account to remain unprivileged.

BPs could create a new account, say `eosio.regram`, whose active authority is also unilaterally satisfied by the `eosio.unregd@eosio.code` permission level. Then instead of issuing the necessary EOS within the action, the contract can simply `buyram` using the `eosio.regram` account. BPs would first need to load up the `eosio.regram` account with sufficient EOS, likely by issuing it or maybe pulling it from existing accumulated fees (I wonder if it needs referendum approval either way?). 

But by doing it in this indirect way, if something went wrong with the contract and it calculated a very large amount of EOS to use to buy the RAM, the damage would be limited to the amount of EOS kept in the `eosio.regram` account. 

The contract could be made even more robust with a further safeguard which adds an assertion within the `unregd::regaccount` action to limit how large the `amount_to_purchase_8kb_of_RAM` value can be. Another action could be added to the contract (that requires the authority of `_self`) to change that limit, so that BPs could explicitly adjust it as necessary without needing to update the whole contract.